packages/java/endpoint/src/main/java/dev/hilla/internal/hotswap/EndpointHotSwapListener.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2000-2023 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package dev.hilla.internal.hotswap;
+
+import com.vaadin.flow.internal.BrowserLiveReload;
+import dev.hilla.EndpointController;
+import dev.hilla.engine.EngineConfiguration;
+import dev.hilla.engine.GeneratorProcessor;
+import dev.hilla.engine.ParserProcessor;
+import jakarta.annotation.PostConstruct;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Optional;
+
+class EndpointHotSwapListener implements HotSwapListener {
+
+    private static final Logger LOGGER = LoggerFactory
+            .getLogger(EndpointHotSwapListener.class);
+
+    private final EndpointController endpointController;
+
+    private final EndpointHotSwapService endpointHotSwapService;
+
+    public EndpointHotSwapListener(EndpointController endpointController,
+            EndpointHotSwapService endpointHotSwapService) {
+        this.endpointController = endpointController;
+        this.endpointHotSwapService = endpointHotSwapService;
+    }
+
+    @PostConstruct
+    private void registerForHotSwapChanges() {
+        endpointHotSwapService.addHotSwapListener(this);
+    }
+
+    @Override
+    public void endpointChanged(EndpointChangedEvent event) {
+        EngineConfiguration engineConfiguration;
+        try {
+            engineConfiguration = EngineConfiguration
+                    .loadDirectory(event.buildDir());
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        ParserProcessor parser = new ParserProcessor(engineConfiguration,
+                this.getClass().getClassLoader());
+        parser.process();
+        GeneratorProcessor generator = new GeneratorProcessor(
+                engineConfiguration, "node");
+        generator.process();
+
+        this.endpointController.registerEndpoints();
+        reload(event.browserLiveReload());
```

Comment:
Why should you reload the browser if an endpoint changes?

Reply:
Vite watches the frontend files and will do hot replace or reload or whatever is needed for that

Reply:
IMO, It is more convenient for the developer to see where in view's code might be broken due to the changes right away, rather than start entering data and clicking button to see errors in the browser's console. 

Reloading the browser automatically triggers the clientside devtools to show the error in the view code with nice styles and has no bad effects on anything (to my understanding). Do you see a pitfall cause by reloading the page?

Reply:
Why does not Vite handle this?

Reply:
I was experiencing some unreliability when Vite wasn't conveniently hot-reloading the page (as it supposed to). However, at that time, I tried this with FSWatch service to regenerate the TS files, so there was a chance of things happening not in the same order as I expected... I'll give it a try to remove the BrowserReload and see what happens.

Reply:
BTW, I had to disable the Vaadin HotSwap plugin since it is reloading the browser with a different timing and in a separate thread that we need to in Hilla.

Reply:
> Why should you reload the browser if an endpoint changes?

What is the harm in doing that? If no significant harm is made, why bother?

Reply:
I disabled the browser reload and Vite is not triggering the hot-replace/reload. 
Currently, I don't know why it isn't working, since the generated files have already changed under `frontend/generated` directory, and even the IDE is detecting the change (e.g. renamed endpoint method). I suggest to ignore this for now and improve it later.